### DISCUSSION TITLE
Unit change of each files in config

### DISCUSSION THREAD

**Discussion by Deba57** [dtcenter/METplus#1007] [Configuration] [answered]
Hi,
I am using MET 8 series stats to verify precipitation rate between a model
and observation for 2 months (240 files for both). I realized I need to
multiply each of the obs prate by a factor to match the units with obs.
How and where do I specify that in Configure file used for this case ?
Thanks
Debasish

**Comment by j-opatz** [comment, marked as answer]
Hi Debasish, and thank you for your question. There are currently two ways you can modify your forecast or observation fields by a set value: the first is within your configuration file using `convert`. Take a look at the MET User's Guide entry for `convert`, which is located under the [forecast dictionary entry](https://met.readthedocs.io/en/develop/Users_Guide/config_options.html#fcst) (but can be applied to observations as well). To apply a constant factor change of 0.15 to your forecast field with `convert`, it might look something like 

fcst = {
   convert(x) = x*0.15;
   field = [ { name  = "var1"; level = "(*,*)";} ];
   }

The second method would be to use a Python script and use the Python embedding ability of MET to read those altered fields in. This is a good idea if you need to do complex factor changes, i.e. changing a forecast field based on the paired observation field. More information on this topic is in Chapter [34](https://met.readthedocs.io/en/develop/Users_Guide/appendixF.html#appendix-f-python-embedding) of the User's Guide.

**Reply by Deba57** [reply]
Thanks for reply. I found 
convert(x) = x0.15; syntax does not work, but 
convert(x) = 0.15x; worked and assuming it will do for all the files in that list for Fcst (I have more than 200 in that list and levels are different). Please confirm

**Reply by Deba57** [reply]
Also, one more thing, if I want to compute Gridstat from a forecast and obs file pairs for any variable but want to compute it over ocean only, but the landsea mask resides in a different file (other than fcst and obs that are listed in gridstat), how do I bring that information to compute statistics? 
thanks
Debasish

**Reply by JohnHalleyGotway** [reply]
@Deba57 and @j-opatz I wanted to jump in here on the details. I was surprised to see the syntax listed as... "convert(x)=x0.15;"... with no star character to indicate multiplication. But then I just realized that it's an artifact of the GitHub editor. It doesn't render star characters when typed in the main body. However, if you format it as "code" it renders well. You can insert a code block using 3 back ticks to begin it and 3 back ticks to end it. Like this:
```
convert(x) = x * 0.15;
```

**Reply by Deba57** [reply]
Thanks @JohnHalleyGotway  and @j-opatz. Let mw know your solution for using mask file mentioned above.

**Reply by JohnHalleyGotway** [reply]
@Deba57 having the land/sea mask data in a separate file is fine. You should be able to configure the mask.poly entry in the Grid-Stat config file to point to it. I'd recommend running it through gen_vx_mask first, like this:
```
gen_vx_mask \
sample_input_data.grib sample_input_data.grib water_mask.nc \
-type data -mask_field 'name="LAND"; level="L0";' -thresh eq0 -name WATER
```
Above I just listed your input file twice and picked an output mask file name. We're doing "data" masking using the field I specified there. In GRIB, land/sea masks are often named LAND and encode water as a value of 0. If you data is named differently, just adjust this configuration.

And then just list the full path to the mask file in the mask.poly of the Grid-Stat configuration file, like this:
```
mask = {
   grid = [ ];
   poly = [ "/full/path/to/water_mask.nc" ];
}
```

**Reply by Deba57** [reply]
Thanks John. I am able to create mask file using your example above. But when I use it in gridstat it shows the following error 
DEBUG 3: Use the observation grid.
DEBUG 3: Grid Definition: Projection: Lat/Lon Nx: 360 Ny: 181 lat_ll: -90.000 lon_ll: -0.000 delta_lat: 1.000 delta_lon: 1.000
DEBUG 2: Processing masking regions.
WARNING:
WARNING: Dictionary::lookup_string() -> string lookup failed for name "mask.poly"
WARNING:
ERROR  :
ERROR  : Dictionary::lookup_string() -> lookup failed for name "mask.poly"

In the config file I used this 
mask = {
   grid = [  ];
   poly = [ "/work1/proj1/temp/water_mask.nc" ];
};

filedump of water_mask.nc shows following : 

Variable: f
Type: file
filename:       water_mask
path:   work1/proj1/temp/water_mask.nc
Number of Attributes:10
    FileOrigins :       File water_mask.nc generated 20210719_223438 UTC on host t45w2 by the MET gen_vx_mask tool
    MET_version :       V8.0
    MET_tool    :       gen_vx_mask
    Projection  :       LatLon
    lat_ll      :       -90.000000 degrees_north
    lon_ll      :       0.000000 degrees_east
    delta_lat   :       1.000000 degrees
    delta_lon   :       1.000000 degrees
    Nlat        :       181 grid_points
    Nlon        :       360 grid_points

dimensions:
    lat = 181
    lon = 360

variables:
    Variable: lat
    Type: float
    Total Size: 181 values
                724 bytes
    Number of Dimensions: 1
    Dimensions and sizes:       [ 181 <lat> ]
    Coordinates:
                lat: [-90..90]
        Number of Attributes:        3
            long_name   :       latitude
            units       :       degrees_north
            standard_name       :       latitude

    Variable: lon
    Type: float
    Total Size: 360 values
                1440 bytes
    Number of Dimensions: 1
    Dimensions and sizes:       [ 360 <lon> ]
    Coordinates:
                lon: [ 0..359]
        Number of Attributes:        3
            long_name   :       longitude
            units       :       degrees_east
            standard_name       :       longitude

    Variable: WATER
    Type: float
    Total Size: 65160 values
                260640 bytes
    Number of Dimensions: 2
    Dimensions and sizes:       [ 181 <lat> x 360 <lon> ]
    Coordinates:
                lat: [-90..90]
                lon: [ 0..359]
        Number of Attributes:        3
            long_name   :       WATER masking region
            _FillValue  :       -9999
            mask_type   :       data==0


            Any help on this. Thanks

**Reply by JohnHalleyGotway** [reply]
Hmm, that's an odd error message:
```
ERROR : Dictionary::lookup_string() -> lookup failed for name "mask.poly"
```
It means that Grid-Stat is having trouble parsing that string from your configuration file for some reason. I don't think it has anything to do with the format of that mask file at all. Perhaps there's just some odd non-printable characters included in that file, which is more likely if you're using Windows? Here's what I'd recommend:
(1) Get back to a state where Grid-Stat worked by setting:
```
mask = {
   grid = [ "FULL" ];
   poly = [ ];
};
```
(2) Run Grid-Stat to make sure that it runs without error.
(3) Edit your config file by adding back in that path:
```
mask = {
   grid = [ "FULL" ];
   poly = [ "/work1/proj1/temp/water_mask.nc" ];
};
```
(4) Re-run Grid-Stat to see if it works this time.

If not, please attach your actual config file to this discussion. I suspect you'll need to add a ".txt" suffix to it for GitHub to allow you to attach it. I can grab it and look for any odd, non-printable characters that are wreaking havoc.
